File: src/Tests/Performance/PerfChannel/PerfChannel.go
package main

import (
	"fmt"
	"time"
)

// Producer/consumer over a buffered channel with one goroutine: measures
// channel send/receive and goroutine scheduling (channel<T> emulation in C#).

func run(n int) int {
	ch := make(chan int, 128)
	done := make(chan int)

	go func() {
		total := 0
		for v := range ch {
			total += v & 4095
		}
		done <- total
	}()

	for i := 0; i < n; i++ {
		ch <- i
	}
	close(ch)

	return <-done
}

func main() {
	start := time.Now().UnixNano()

	total := run(1000000)

	elapsed := time.Now().UnixNano() - start
	fmt.Println("checksum:", total)
	fmt.Println("elapsed_ns:", elapsed)
}

File: src/Tests/Performance/PerfChannel/PerfChannel.cs
namespace go;

using fmt = fmt_package;
using time = time_package;

partial class main_package {

internal static nint run(nint n) {
    var ch = new channel<nint>(128);
    var done = new channel<nint>(1);
    var chʗ1 = ch;
    var doneʗ1 = done;
    goǃ(() => {
        nint total = 0;
        foreach (var v in chʗ1) {
            total += (nint)(v & 4095);
        }
        doneʗ1.ᐸꟷ(total);
    });
    for (nint i = 0; i < n; i++) {
        ch.ᐸꟷ(i);
    }
    close(ch);
    return ᐸꟷ(done);
}

internal static void Main() {
    var start = time.Now().UnixNano();
    nint total = run(1000000);
    var elapsed = time.Now().UnixNano() - start;
    fmt.Println("checksum:", total);
    fmt.Println("elapsed_ns:", elapsed);
}

} // end main_package
